testing/src/scenario/state.py

```diff
@@ -212,20 +234,22 @@ class JujuLogLine:
 
 
 @dataclasses.dataclass(frozen=True)
+@_detach_dicts_and_lists
```

Comment:
Not sure, but what do you think about just adding `__post_init__` on these explicitly? It's 2 lines (plus a blank line) instead of 1, but perhaps clearer (and we wouldn't need to call the original `__post_init__`, as there isn't one, or we'd be adding to the existing one):

```python
def __post_init__(self):
    _deepcopy_mutable_fields(self)
```